# Project: WeRateDogs Twitter Archive

## Table of Contents
<ul>
    <li><a href="#intro">Introduction</a></li>
    <li><a href="#wrangle">Data Wrangling</a></li>
    <li><a href="#analyze">Data Analysis</a></li>
    <li><a href="#visualize">Data Visualization</a></li>
    <li><a href="#conclusion">Conclusion</a></li>
</ul>

## Introduction
<a id='intro'></a>

> text

> text

In [1]:
# import libraries
import glob
import json
import os
import pandas as pd
import requests
import tweepy

## Data Wrangling
<a id='wrangle'></a>

### Gathering

In [2]:
# read in twitter-archive-enhanced file
df1 = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
# get response object of info from url
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

In [4]:
# create tsv file for image predictions if it doesn't already exist
if not os.path.exists('image-predictions.tsv'):
    with open(os.path.join(url.split('/')[-1]), mode = 'wb') as file:
        file.write(response.content)

In [5]:
# read image predictions into pandas DataFrame
df2 = pd.read_csv('image-predictions.tsv', delimiter='\t', encoding='utf-8')

In [6]:
# use key, token, and secrets to access Twitter API
# REMOVE KEYS WHEN SUBMITTING!!!


# use tweepy to access api
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)
api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

> Althought the following section could have been done simpler by pulling the three columns of information from the API directly into the df3_list and then converting it into a DataFrame, it was stated in the project details that the tweets need to be stored in .txt files and then read into the DataFrame.

In [7]:
# make folder to store tweet info if it doesn't already exist
folder_name = 'tweet_infos'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)

In [8]:
# create error list to catch tweet_ids where the tweet_info doesn't get pulled
error_list = []
# go through tweet_ids in df1 and saving tweet_info as txt files in tweet_infos folder
for tweet_id in df1.tweet_id:
    try:
        tweet_info = json.dumps(api.get_status(str(tweet_id), tweet_mode='extended')._json)
        with open(os.path.join(folder_name, '{}{}'.format(tweet_id,'.txt')), mode='w') as file:
            file.write(tweet_info)
            print(tweet_id)
    except Exception as e:
        print('error')
        error_list.append(tweet_id)

892420643555336193
892177421306343426
891815181378084864
891689557279858688
891327558926688256
891087950875897856
890971913173991426
890729181411237888
890609185150312448
890240255349198849
890006608113172480
889880896479866881
889665388333682689
889638837579907072
889531135344209921
889278841981685760
888917238123831296
888804989199671297
888554962724278272
error
888078434458587136
887705289381826560
887517139158093824
887473957103951883
887343217045368832
887101392804085760
886983233522544640
886736880519319552
886680336477933568
886366144734445568
886267009285017600
886258384151887873
886054160059072513
885984800019947520
885528943205470208
885518971528720385
885311592912609280
885167619883638784
884925521741709313
884876753390489601
884562892145688576
884441805382717440
884247878851493888
884162670584377345
883838122936631299
883482846933004288
883360690899218434
883117836046086144
882992080364220416
882762694511734784
882627270321602560
882268110199369728
882045870035918850
881906

819347104292290561
819238181065359361
819227688460238848
819015337530290176
819015331746349057
819006400881917954
819004803107983360
818646164899774465
818627210458333184
818614493328580609
818588835076603904
818536468981415936
818307523543449600
818259473185828864
818145370475810820
817908911860748288
817827839487737858
817777686764523521
817536400337801217
817502432452313088
817423860136083457
817415592588222464
817181837579653120
817171292965273600
817120970343411712
817056546584727552
816829038950027264
816816676327063552
816697700272001025
816450570814898180
816336735214911488
816091915477250048
816062466425819140
816014286006976512
815990720817401858
815966073409433600
815745968457060357
815736392542261248
815639385530101762
815390420867969024
814986499976527872
814638523311648768
814578408554463233
814530161257443328
814153002265309185
813944609378369540
813910438903693312
813812741911748608
813800681631023104
813217897535406080
813202720496779264
813187593374461952
813172488309

760656994973933572
760641137271070720
760539183865880579
760521673607086080
760290219849637889
760252756032651264
760190180481531904
error
759943073749200896
759923798737051648
759846353224826880
759793422261743616
759566828574212096
759557299618865152
759447681597108224
759446261539934208
759197388317847553
759159934323924993
759099523532779520
759047813560868866
758854675097526272
758828659922702336


Rate limit reached. Sleeping for: 684


758740312047005698
758474966123810816
758467244762497024
758405701903519748
758355060040593408
758099635764359168
758041019896193024
757741869644341248
757729163776290825
757725642876129280
757611664640446465
757597904299253760
757596066325864448
757400162377592832
757393109802180609
757354760399941633
756998049151549440
756939218950160384
756651752796094464
756526248105566208
756303284449767430
756288534030475264
756275833623502848
755955933503782912
755206590534418437
755110668769038337
754874841593970688
754856583969079297
754747087846248448
754482103782404096
754449512966619136
754120377874386944
error
753655901052166144
753420520834629632
753398408988139520
753375668877008896
753298634498793472
753294487569522689
753039830821511168
753026973505581056
752932432744185856
752917284578922496
752701944171524096
752682090207055872
752660715232722944
752568224206688256
752519690950500352
752334515931054080
752309394570878976
752173152931807232
751950017322246144
751937170840121344
751830

705475953783398401
705442520700944385
705428427625635840
705239209544720384
705223444686888960
705102439679201280
705066031337840642
704871453724954624
704859558691414016
704847917308362754
704819833553219584
704761120771465216
704499785726889984
704491224099647488
704480331685040129
704364645503647744
704347321748819968
704134088924532736
704113298707505153
704054845121142784
703774238772166656
703769065844768768
703631701117943808
703611486317502464
703425003149250560
703407252292673536
703382836347330562
703356393781329922
703268521220972544
703079050210877440
703041949650034688
702932127499816960
702899151802126337
702684942141153280
702671118226825216
702598099714314240
702539513671897089
702332542343577600
702321140488925184
702276748847800320
702217446468493312
701981390485725185
701952816642965504
701889187134500865
701805642395348998
701601587219795968
701570477911896070
701545186879471618
701214700881756160
700890391244103680
700864154249383937
700847567345688576
700796979434

678424312106393600
678410210315247616
678399652199309312
678396796259975168
678389028614488064
678380236862578688
678341075375947776
678334497360859136
678278586130948096
678255464182861824
678023323247357953
678021115718029313
677961670166224897
677918531514703872
677895101218201600
677716515794329600
677700003327029250
677698403548192770
677687604918272002
677673981332312066
677662372920729601
677644091929329666
677573743309385728
677565715327688705
677557565589463040
677547928504967168
677530072887205888
677335745548390400
677334615166730240
677331501395156992
677328882937298944
677314812125323265
677301033169788928
677269281705472000
677228873407442944


Rate limit reached. Sleeping for: 683


677187300187611136
676975532580409345
676957860086095872
676949632774234114
676948236477857792
676946864479084545
676942428000112642
676936541936185344
676916996760600576
676897532954456065
676864501615042560
676821958043033607
676819651066732545
676811746707918848
676776431406465024
676617503762681856
676613908052996102
676606785097199616
676603393314578432
676593408224403456
676590572941893632
676588346097852417
676582956622721024
676575501977128964
676533798876651520
676496375194980353
676470639084101634
676440007570247681
676430933382295552
676263575653122048
676237365392908289
676219687039057920
676215927814406144
676191832485810177
676146341966438401
676121918416756736
676101918813499392
676098748976615425
676089483918516224
675898130735476737
675891555769696257
675888385639251968
675878199931371520
675870721063669760
675853064436391936
675849018447167488
675845657354215424
675822767435051008
675820929667219457
675798442703122432
675781562965868544
675740360753160193
675710890956

668226093875376128
668221241640230912
668204964695683073
668190681446379520
668171859951755264
668154635664932864
668142349051129856
668113020489474048
667937095915278337
667924896115245057
667915453470232577
667911425562669056
667902449697558528
667886921285246976
667885044254572545
667878741721415682
667873844930215936
667866724293877760
667861340749471744
667832474953625600
667806454573760512
667801013445750784
667793409583771648
667782464991965184
667773195014021121
667766675769573376
667728196545200128
667724302356258817
667550904950915073
667550882905632768
667549055577362432
667546741521195010
667544320556335104
667538891197542400
667534815156183040
667530908589760512
667524857454854144
667517642048163840
667509364010450944
667502640335572993
667495797102141441
667491009379606528
667470559035432960
667455448082227200
667453023279554560
667443425659232256
667437278097252352
667435689202614272
667405339315146752
667393430834667520
667369227918143488
667211855547486208
667200525029

In [9]:
# create list to append information into
df3_list = []
# fill list with info from text files
for tweet_info in glob.glob('tweet_infos/*.txt'):
    with open(tweet_info, encoding='utf-8') as file:
        content = json.load(file)
        tweet_id = content['id']
        retweet_count = content['retweet_count']            
        favorite_count = content['favorite_count']
        df3_list.append({'tweet_id' : tweet_id,
                   'retweet_count' : retweet_count,
                   'favorite_count' : favorite_count})
        print(tweet_id)

666020888022790149
666029285002620928
666033412701032449
666044226329800704
666049248165822465
666050758794694657
666051853826850816
666055525042405380
666057090499244032
666058600524156928
666063827256086533
666071193221509120
666073100786774016
666082916733198337
666094000022159362
666099513787052032
666102155909144576
666104133288665088
666268910803644416
666273097616637952
666287406224695296
666293911632134144
666337882303524864
666345417576210432
666353288456101888
666362758909284353
666373753744588802
666396247373291520
666407126856765440
666411507551481857
666418789513326592
666421158376562688
666428276349472768
666430724426358785
666435652385423360
666437273139982337
666447344410484738
666454714377183233
666644823164719104
666649482315059201
666691418707132416
666701168228331520
666739327293083650
666776908487630848
666781792255496192
666786068205871104
666804364988780544
666817836334096384
666826780179869698
666835007768551424
666837028449972224
666983947667116034
666996132027

674754018082705410
674764817387900928
674767892831932416
674774481756377088
674781762103414784
674788554665512960
674790488185167872
674793399141146624
674800520222154752
674805413498527744
674999807681908736
675003128568291329
675006312288268288
675015141583413248
675047298674663426
675109292475830276
675111688094527488
675113801096802304
675135153782571009
675145476954566656
675146535592706048
675147105808306176
675149409102012420
675153376133427200
675166823650848770
675334060156301312
675349384339542016
675354435921575936
675362609739206656
675372240448454658
675432746517426176
675483430902214656
675489971617296384
675497103322386432
675501075957489664
675517828909424640
675522403582218240
675531475945709568
675534494439489536
675706639471788032
675707330206547968
675710890956750848
675740360753160193
675781562965868544
675798442703122432
675820929667219457
675822767435051008
675845657354215424
675849018447167488
675853064436391936
675870721063669760
675878199931371520
675888385639

697516214579523584
697575480820686848
697596423848730625
697616773278015490
697630435728322560
697881462549430272
697943111201378304
697990423684476929
697995514407682048
698178924120031232
698195409219559425
698262614669991936
698342080612007937
698355670425473025
698549713696649216
698635131305795584
698703483621523456
698710712454139905
698907974262222848
698953797952008193
698989035503689728
699036661657767936
699060279947165696
699072405256409088
699079609774645248
699088579889332224
699323444782047232
699370870310113280
699413908797464576
699423671849451520
699434518667751424
699446877801091073
699691744225525762
699775878809702401
699779630832685056
699788877217865730
699801817392291840
700002074055016451
700029284593901568
700062718104104960
700143752053182464
700151421916807169
700167517596164096
700462010979500032
700505138482569216
700518061187723268
700747788515020802
700796979434098688
700847567345688576
700864154249383937
700890391244103680
701214700881756160
701545186879

750026558547456000
750041628174217216
750056684286914561
750071704093859840
750086836815486976
750101899009982464
750117059602808832
750132105863102464
750147208377409536
750381685133418496
750383411068534784
750429297815552001
750506206503038976
750719632563142656
750868782890057730
751132876104687617
751205363882532864
751251247299190784
751456908746354688
751538714308972544
751583847268179968
751598357617971201
751793661361422336
751830394383790080
751937170840121344
751950017322246144
752173152931807232
752309394570878976
752334515931054080
752519690950500352
752568224206688256
752660715232722944
752682090207055872
752701944171524096
752917284578922496
752932432744185856
753026973505581056
753039830821511168
753294487569522689
753298634498793472
753375668877008896
753398408988139520
753420520834629632
753655901052166144
754120377874386944
754449512966619136
754482103782404096
754747087846248448
754856583969079297
754874841593970688
755110668769038337
755206590534418437
755955933503

809808892968534016
809920764300447744
810254108431155201
810284430598270976
810657578271330305
810896069567610880
810984652412424192
811386762094317568
811627233043480576
811647686436880384
811744202451197953
811985624773361665
812372279581671427
812466873996607488
812503143955202048
812709060537683968
812781120811126785
813051746834595840
813066809284972545
813081950185472002
813096984823349248
813112105746448384
813127251579564032
813130366689148928
813142292504645637
813157409116065792
813172488309972993
813187593374461952
813202720496779264
813217897535406080
813800681631023104
813812741911748608
813910438903693312
813944609378369540
814153002265309185
814530161257443328
814578408554463233
814638523311648768
814986499976527872
815390420867969024
815639385530101762
815736392542261248
815745968457060357
815966073409433600
815990720817401858
816014286006976512
816062466425819140
816091915477250048
816336735214911488
816450570814898180
816697700272001025
816816676327063552
816829038950

883117836046086144
883360690899218434
883482846933004288
883838122936631299
884162670584377345
884247878851493888
884441805382717440
884562892145688576
884876753390489601
884925521741709313
885167619883638784
885311592912609280
885518971528720385
885528943205470208
885984800019947520
886054160059072513
886258384151887873
886267009285017600
886366144734445568
886680336477933568
886736880519319552
886983233522544640
887101392804085760
887343217045368832
887473957103951883
887517139158093824
887705289381826560
888078434458587136
888554962724278272
888804989199671297
888917238123831296
889278841981685760
889531135344209921
889638837579907072
889665388333682689
889880896479866881
890006608113172480
890240255349198849
890609185150312448
890729181411237888
890971913173991426
891087950875897856
891327558926688256
891689557279858688
891815181378084864
892177421306343426
892420643555336193


In [10]:
# convert list to DataFrame
df3 = pd.DataFrame(df3_list, columns = ['tweet_id', 'retweet_count', 'favorite_count'])

### Assessing

In [54]:
df1.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [12]:
df1.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [13]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

`df1` columns:
* **tweet_id**: the 18 digit unique id for each tweet in the provided WeRateDogs twitter archive
* **in_reply_to_status_id**: if a reply tweet, the 21 digit status id of the original tweet
* **in_reply_to_user_id**: if a reply tweet, the user id of the original tweet
* **timestamp**: the time the tweet was recorded
* **source**: the reference website the tweet was sent from (Twitter for iPhone, Twitter Web Client, Vine, or TweetDeck)
* **text**: the full text of the tweet including shortened link
* **retweeted_status_id**: if a retweet, the 21 digit status id of the original tweet
* **retweeted_status_user_id**: if a retweet, the user id of the original tweet
* **retweeted_status_timestamp**: if a retweet, the timestamp of the original tweet
* **expanded_urls**: the full url of the tweet
* **rating_numerator**: the numerator for the rating extracted from the text of the tweet
* **rating_denominator**: the denominator of the rating extracted from the text of the tweet
* **name**: the name of the dog extracted from the text of the tweet
* **doggo**: either 'none' or 'doggo' if 'doggo' is used in the text of the tweet
* **floofer**:either 'none' or 'floofer' if 'floofer' is used in the text of the tweet
* **pupper**:either 'none' or 'pupper' if 'pupper' is used in the text of the tweet
* **puppo**:either 'none' or 'puppo' if 'puppo' is used in the text of the tweet

In [55]:
df2.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [15]:
df2.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [16]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


`df2` columns:
* **tweet_id**: the 18 digit unique id for each tweet in the provided WeRateDogs twitter archive
* **jpg_url**: the url of the image for the tweet
* **img_num**: 1-4 value???
* **p1**: the top dog breed prediction from the neural network
* **p1_conf**: the confidence of the prediction
* **p1_dog**: whether the prediction is a breed of dog
* **p2**: the second dog breed prediction from the neural network
* **p2_conf**: the confidence of the prediction
* **p2_dog**: whether the prediction is a breed of dog
* **p3**: the third dog breed prediction from the neural network
* **p3_conf**: the confidence of the prediction
* **p3_dog**: whether the prediction is a breed of dog|

In [56]:
df3.head()

tweet_id  retweet_count  favorite_count
0  666020888022790149            478            2474
1  666029285002620928             45             122
2  666033412701032449             43             119
3  666044226329800704            135             280
4  666049248165822465             41             101

In [18]:
df3.describe()

tweet_id  retweet_count  favorite_count
count  2.333000e+03    2333.000000     2333.000000
mean   7.419279e+17    2783.389198     7722.854265
std    6.820607e+16    4711.250001    11989.617726
min    6.660209e+17       1.000000        0.000000
25%    6.782786e+17     560.000000     1339.000000
50%    7.184547e+17    1302.000000     3363.000000
75%    7.986654e+17    3238.000000     9479.000000
max    8.924206e+17   79986.000000   160040.000000

In [19]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2333 entries, 0 to 2332
Data columns (total 3 columns):
tweet_id          2333 non-null int64
retweet_count     2333 non-null int64
favorite_count    2333 non-null int64
dtypes: int64(3)
memory usage: 54.8 KB


`df3` columns:
* **tweet_id**: the 18 digit unique id for each tweet in the provided WeRateDogs twitter archive
* **retweet_count**: the number of retweets the tweet received
* **favorite_count**: the number of favorites (likes) the tweet received

#### Quality
1. For df1 the rating_numerator and rating_denominator for tweet_id 835246439529840640 should be 13 and 10, respectively.
2. For df1 some of the names are inputted incorrectly and are random lowercase words from the text.
3. In df1 not all rows labelled with doggo are done so accurately.
4. In df1 not all rows labelled with floofer are done so accurately.
5. In df1 not all rows labelled with pupper are done so accurately.
6. In df1 not all rows labelled with puppo are done so accurately.
7. The df1 rating_numerator max values of 1776, 420, and 666 are specific joke values which are inconsistent with the rest of the data.
8. The df1 rating_numerator minimum value is 0 in two instances which are both accurate to their tweets are inconsistent with the rest of the data.
9. Across dataframes, all ID values should be stored as strings not floats.
10. In df1 the names column has Nan values listed as strings 'None'. 
11. In df1 timestamp and retweeted_status_timestamp should be in datetime format

#### Tidiness
1. In df1 the in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, and expanded_urls columns can be removed. 
2. In df1 can collapse doggo, floofer, pupper, and puppo columns into one column.
3. df1 can store all three dataframes without sacrificing tidiness standards.

In [ ]:
df1.tweet_id[0]

In [ ]:
df1.expanded_urls[0]

In [ ]:
int(df1.expanded_urls[0].split('/')[-3])

In [ ]:
list1 = df1[df1['doggo'] == 'doggo'].tweet_id

In [ ]:
df1[df1['text'].str.contains('doggo')]

In [ ]:
pd.set_option('display.max_colwidth', -1)
df1[df1['tweet_id'] == 876537666061221889]['text']

In [ ]:
df1[df1['rating_numerator'] == 0]

In [ ]:
api.get_status(835246439529840640).text

In [ ]:
pd.set_option("display.max_rows", 99999)
df1['name'].value_counts()

In [ ]:
not_name_id = []
pd.set_option('display.max_colwidth', -1)
for index, row in df1.iterrows():
    if row['name'].islower() == True:
        not_name.append(row['tweet_id'])
        print(row['tweet_id'])
        print(row['text'])

In [32]:
df1[df1['tweet_id'].isin(error_list)]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257                    NaN                  NaN   
95    873697596434513921                    NaN                  NaN   
101   872668790621863937                    NaN                  NaN   
104   872261713294495745                    NaN                  NaN   
118   869988702071779329                    NaN                  NaN   
132   866816280283807744                    NaN                  NaN   
155   861769973181624320                    NaN                  NaN   
182   856602993587888130                    NaN                  NaN   
211   851953902622658560                    NaN                  NaN   
247   845459076796616705                    NaN                  NaN   
253   844704788403113984                    NaN                  NaN   
260   842892208864923648                    NaN                  NaN   
296   837366284874571778                    NaN                  NaN   
298   837012587749474308                    NaN                  NaN   
363   829374341691346946                    NaN                  NaN   
382   827228250799742977                    NaN                  NaN   
506   812747805718642688                    NaN                  NaN   
566   802247111496568832                    NaN                  NaN   
750   779123168116150273                    NaN                  NaN   
784   775096608509886464                    NaN                  NaN   
818   770743923962707968                    NaN                  NaN   
885   760153949710192640                    NaN                  NaN   
932   754011816964026368                    NaN                  NaN   
1726  680055455951884288                    NaN                  NaN   

                      timestamp  \
19    2017-07-21 01:02:36 +0000   
95    2017-06-11 00:25:14 +0000   
101   2017-06-08 04:17:07 +0000   
104   2017-06-07 01:19:32 +0000   
118   2017-05-31 18:47:24 +0000   
132   2017-05-23 00:41:20 +0000   
155   2017-05-09 02:29:07 +0000   
182   2017-04-24 20:17:23 +0000   
211   2017-04-12 00:23:33 +0000   
247   2017-03-25 02:15:26 +0000   
253   2017-03-23 00:18:10 +0000   
260   2017-03-18 00:15:37 +0000   
296   2017-03-02 18:17:34 +0000   
298   2017-03-01 18:52:06 +0000   
363   2017-02-08 17:00:26 +0000   
382   2017-02-02 18:52:38 +0000   
506   2016-12-24 19:52:31 +0000   
566   2016-11-25 20:26:31 +0000   
750   2016-09-23 01:00:13 +0000   
784   2016-09-11 22:20:06 +0000   
818   2016-08-30 22:04:05 +0000   
885   2016-08-01 16:43:19 +0000   
932   2016-07-15 17:56:40 +0000   
1726  2015-12-24 16:00:30 +0000   

                                                 source  \
19    <a href="http://twitter.com/download/iphone" r...   
95    <a href="http://twitter.com/download/iphone" r...   
101   <a href="http://twitter.com/download/iphone" r...   
104   <a href="http://twitter.com/download/iphone" r...   
118   <a href="http://twitter.com/download/iphone" r...   
132   <a href="http://twitter.com/download/iphone" r...   
155   <a href="http://twitter.com/download/iphone" r...   
182   <a href="http://twitter.com/download/iphone" r...   
211   <a href="http://twitter.com/download/iphone" r...   
247   <a href="http://twitter.com/download/iphone" r...   
253   <a href="http://twitter.com/download/iphone" r...   
260   <a href="http://twitter.com/download/iphone" r...   
296   <a href="http://twitter.com/download/iphone" r...   
298   <a href="http://twitter.com/download/iphone" r...   
363   <a href="http://twitter.com/download/iphone" r...   
382   <a href="http://twitter.com/download/iphone" r...   
506   <a href="http://twitter.com/download/iphone" r...   
566   <a href="http://twitter.com/download/iphone" r...   
750   <a href="http://twitter.com/download/iphone" r...   
784   <a href="http://twitter.com/download/iphone" r...   
818   <a href="http://twitter.com/download/iphone" r...   
885   <a href="http://twitter.com/downloa

In [38]:
df1[df1['name'] == 'None']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
5     891087950875897856                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
30    886267009285017600           8.862664e+17         2.281182e+09   
32    886054160059072513                    NaN                  NaN   
35    885518971528720385                    NaN                  NaN   
37    885167619883638784                    NaN                  NaN   
41    884441805382717440                    NaN                  NaN   
42    884247878851493888                    NaN                  NaN   
47    883117836046086144                    NaN                  NaN   
55    881633300179243008           8.816070e+17         4.738443e+07   
59    880872448815771648                    NaN                  NaN   
62    880095782870896641                    NaN                  NaN   
64    879674319642796034           8.795538e+17         3.105441e+09   
72    878604707211726852                    NaN                  NaN   
78    877611172832227328                    NaN                  NaN   
83    876537666061221889                    NaN                  NaN   
88    875097192612077568                    NaN                  NaN   
89    875021211251597312                    NaN                  NaN   
93    874057562936811520                    NaN                  NaN   
96    873580283840344065                    NaN                  NaN   
99    872967104147763200                    NaN                  NaN   
100   872820683541237760                    NaN                  NaN   
101   872668790621863937                    NaN                  NaN   
103   872486979161796608                    NaN                  NaN   
110   871102520638267392                    NaN                  NaN   
112   870804317367881728                    NaN                  NaN   
113   870726314365509632           8.707262e+17         1.648776e+07   
...                  ...                    ...                  ...   
2294  667138269671505920                    NaN                  NaN   
2298  667070482143944705           6.670655e+17         4.196984e+09   
2299  667065535570550784                    NaN                  NaN   
2301  667044094246576128                    NaN                  NaN   
2305  666837028449972224                    NaN                  NaN   
2306  666835007768551424                    NaN                  NaN   
2307  666826780179869698                    NaN                  NaN   
2310  666786068205871104                    NaN                  NaN   
2316  666649482315059201                    NaN                  NaN   
2320  666437273139982337                    NaN                  NaN   
2321  666435652385423360                    NaN                  NaN   
2322  666430724426358785                    NaN                  NaN   
2323  666428276349472768                    NaN                  NaN   
2324  666421158376562688                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN         

### Cleaning

In [57]:
# copy dataframes for cleaning
df1_clean = df1.copy()
df2_clean = df2.copy()
df3_clean = df3.copy()

#### Issue

In df1 the in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, and expanded_urls columns can be removed. 

#### Define

Use dataframe.drop to remove unnecessary columns.

#### Code

In [59]:
df1_clean.drop(columns=['in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 'expanded_urls'])

tweet_id                  timestamp  \
0     892420643555336193  2017-08-01 16:23:56 +0000   
1     892177421306343426  2017-08-01 00:17:27 +0000   
2     891815181378084864  2017-07-31 00:18:03 +0000   
3     891689557279858688  2017-07-30 15:58:51 +0000   
4     891327558926688256  2017-07-29 16:00:24 +0000   
5     891087950875897856  2017-07-29 00:08:17 +0000   
6     890971913173991426  2017-07-28 16:27:12 +0000   
7     890729181411237888  2017-07-28 00:22:40 +0000   
8     890609185150312448  2017-07-27 16:25:51 +0000   
9     890240255349198849  2017-07-26 15:59:51 +0000   
10    890006608113172480  2017-07-26 00:31:25 +0000   
11    889880896479866881  2017-07-25 16:11:53 +0000   
12    889665388333682689  2017-07-25 01:55:32 +0000   
13    889638837579907072  2017-07-25 00:10:02 +0000   
14    889531135344209921  2017-07-24 17:02:04 +0000   
15    889278841981685760  2017-07-24 00:19:32 +0000   
16    888917238123831296  2017-07-23 00:22:39 +0000   
17    888804989199671297  2017-07-22 16:56:37 +0000   
18    888554962724278272  2017-07-22 00:23:06 +0000   
19    888202515573088257  2017-07-21 01:02:36 +0000   
20    888078434458587136  2017-07-20 16:49:33 +0000   
21    887705289381826560  2017-07-19 16:06:48 +0000   
22    887517139158093824  2017-07-19 03:39:09 +0000   
23    887473957103951883  2017-07-19 00:47:34 +0000   
24    887343217045368832  2017-07-18 16:08:03 +0000   
25    887101392804085760  2017-07-18 00:07:08 +0000   
26    886983233522544640  2017-07-17 16:17:36 +0000   
27    886736880519319552  2017-07-16 23:58:41 +0000   
28    886680336477933568  2017-07-16 20:14:00 +0000   
29    886366144734445568  2017-07-15 23:25:31 +0000   
...                  ...                        ...   
2326  666411507551481857  2015-11-17 00:24:19 +0000   
2327  666407126856765440  2015-11-17 00:06:54 +0000   
2328  666396247373291520  2015-11-16 23:23:41 +0000   
2329  666373753744588802  2015-11-16 21:54:18 +0000   
2330  666362758909284353  2015-11-16 21:10:36 +0000   
2331  666353288456101888  2015-11-16 20:32:58 +0000   
2332  666345417576210432  2015-11-16 20:01:42 +0000   
2333  666337882303524864  2015-11-16 19:31:45 +0000   
2334  666293911632134144  2015-11-16 16:37:02 +0000   
2335  666287406224695296  2015-11-16 16:11:11 +0000   
2336  666273097616637952  2015-11-16 15:14:19 +0000   
2337  666268910803644416  2015-11-16 14:57:41 +0000   
2338  666104133288665088  2015-11-16 04:02:55 +0000   
2339  666102155909144576  2015-11-16 03:55:04 +0000   
2340  666099513787052032  2015-11-16 03:44:34 +0000   
2341  666094000022159362  2015-11-16 03:22:39 +0000   
2342  666082916733198337  2015-11-16 02:38:37 +0000   
2343  666073100786774016  2015-11-16 01:59:36 +0000   
2344  666071193221509120  2015-11-16 01:52:02 +0000   
2345  666063827256086533  2015-11-16 01:22:45 +0000   
2346  666058600524156928  2015-11-16 01:01:59 +0000   
2347  666057090499244032  2015-11-16 00:55:59 +0000   
2348  666055525042405380  2015-11-16 00:49:46 +0000   
2349  666051853826850816  2015-11-16 00:35:11 +0000   
2350  666050758794694657  2015-11-16 00:30:50 +0000   
2351  666049248165822465  2015-11-16 00:24:50 +0000   
2352  666044226329800704  2015-11-16 00:04:52 +0000   
2353  666033412701032449  2015-11-15 23:21:54 +0000   
2354  666029285002620928  2015-11-15 23:05:30 +0000   
2355  666020888022790149  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
5     <a href="http://twitter.com/download/iphone" r...   
6     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
8     <a href="http://twitter.com/download/iphone" r...   
9     <a href="

#### Test

In [60]:
df1_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

#### Issue

In df1 can collapse doggo, floofer, pupper, and puppo columns into one column.

#### Define

#### Code

#### Test

#### Issue

df1 can store all three dataframes without sacrificing tidiness standards.

#### Define

#### Code

#### Test

#### Issue

For df1 the rating_numerator and rating_denominator for tweet_id 835246439529840640 should be 13 and 10, respectively.

#### Define

#### Code

#### Test

#### Issue

For df1 some of the names are inputted incorrectly and are random lowercase words from the text.

#### Define

#### Code

#### Test

#### Issue

In df1 not all rows labelled with doggo are done so accurately.

#### Define

#### Code

#### Test

#### Issue

In df1 not all rows labelled with floofer are done so accurately.

#### Define

#### Code

#### Test

#### Issue

In df1 not all rows labelled with pupper are done so accurately.

#### Define

#### Code

#### Test

#### Issue

In df1 not all rows labelled with puppo are done so accurately.

#### Define

#### Code

#### Test

#### Issue

The df1 rating_numerator max values of 1776, 420, and 666 are specific joke values which are inconsistent with the rest of the data.

#### Define

#### Code

#### Test

#### Issue

The df1 rating_numerator minimum value is 0 in two instances which are both accurate to their tweets are inconsistent with the rest of the data.

#### Define

#### Code

#### Test

#### Issue

Across dataframes, all ID values should be stored as strings not floats.

#### Define

#### Code

#### Test

#### Issue

In df1 the names column has Nan values listed as strings 'None'. 

#### Define

#### Code

#### Test

#### Issue

In df1 timestamp and retweeted_status_timestamp should be in datetime format

#### Define

#### Code

#### Test

## Store

<a id='analyze'></a>
## Data Analysis

<a id='visualize'></a>
## Data Visualization

<a id='conclusion'></a>
## Conclusion

### Data Wrangling Effort

### Data Analyses and Visualizations